mater/lib/src/stores/file.rs

```
@@ -0,0 +1,296 @@
+use std::{io::SeekFrom, path::Path};
+
+use indexmap::IndexMap;
+use ipld_core::cid::Cid;
+use sha2::{Digest, Sha256};
+use tokio::{
+    fs::File,
+    io::{AsyncSeekExt, AsyncWriteExt},
+    sync::{Mutex, RwLock},
+};
+
+use crate::{
+    multicodec::SHA_256_CODE,
+    v1::{self, read_block, write_block},
+    v2::{self},
+    CarV1Header, CarV2Header, Characteristics, Error, Index, IndexEntry, MultihashIndexSorted,
+    SingleWidthIndex,
+};
+
+/// Implements a blockstore that stores blocks in a CARv2 format. Blocks put
+/// into the blockstore can be read back once they are successfully written. The
+/// blocks are written immediately, while the index is stored in memory and
+/// updated incrementally.
+///
+/// The blockstore should be closed once the putting blocks is finished. Upon
+/// closing the blockstore, the index is written out to underlying file.
+pub struct FileBlockstore {
+    // Inner store
+    inner: Mutex<FileBlockstoreInner>,
+    // Index of blocks that will be appended to the file at the finalization
+    index: RwLock<IndexMap<Cid, u64>>,
+}
+
+/// Inner file store. Encapsulating state that is locked and used together.
+struct FileBlockstoreInner {
+    // Car file data store
+    store: File,
+    // The byte length of the CARv1 payload. This is used by the indexing, so we
+    // know the locations of each blocks in the file.
+    data_size: u64,
+    // Is true if the blockstore was finalized.
+    is_finalized: bool,
+}
+
+impl FileBlockstore {
+    /// Create a new blockstore. If file at the path already exists it is truncated.
+    pub async fn new<P>(path: P, roots: Vec<Cid>) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let mut file = File::options()
+            .truncate(true)
```

Why would we override file existing at a given path? Shouldn't we avoid writing to the same path?

I mean, is the CARv2 archive reopenable, or not? If we can't reopen it and append data to it, we should throw an error here and say there is a file.

---

I agree. I was also not entirely sure about truncation.